### PR TITLE
Added PyTorch Sparse Tensor support for `ToSparseTensor`

### PR DESCRIPTION
**Pull request by EdisonLeeeee**

### CODE REVIEW

File: torch_geometric/transforms/to_sparse_tensor.py [hunk 1]
```diff
@@ -4,39 +4,57 @@
...
-from torch_geometric.utils import sort_edge_index
+from torch_geometric.utils import sort_edge_index, to_torch_coo_tensor
 
 
 @functional_transform('to_sparse_tensor')
 class ToSparseTensor(BaseTransform):
     r"""Converts the :obj:`edge_index` attributes of a homogeneous or
     heterogeneous data object into a (transposed)
-    :class:`torch_sparse.SparseTensor` object with key :obj:`adj_t`
+    :class:`torch_sparse.SparseTensor` or a :class:`torch.spase.Tensor`
```
Comment (wsad1) [inline torch_geometric/transforms/to_sparse_tensor.py:14]:
```suggestion
    :class:`torch_sparse.SparseTensor` or a native PyTorch sparse tensor
```
Is `torch.sparse.Tensor ` a new type in pytorch? I thought Pytorch sparse tensor was of type `Tensor`.

Comment (EdisonLeeeee) [reply torch_geometric/transforms/to_sparse_tensor.py:14]: No. Actually `torch.sparse.Tensor` is an alias of `torch.Tensor`.

File: torch_geometric/transforms/to_sparse_tensor.py [hunk 2]
```diff
@@ -4,39 +4,57 @@
...
+    object with key :obj:`adj_t`
     (functional name: :obj:`to_sparse_tensor`).
 
     .. note::
 
         In case of composing multiple transforms, it is best to convert the
-        :obj:`data` object to a :class:`~torch_sparse.SparseTensor` as late as
-        possible, since there exist some transforms that are only able to
+        :obj:`data` object to a :class:`~torch_sparse.SparseTensor` or a
+        :class:`torch.spase.Tensor` as late as possible,
```
Comment (wsad1) [inline torch_geometric/transforms/to_sparse_tensor.py:22]:
```suggestion
        :class:`torch.sparse.Tensor` as late as possible,
```

File: torch_geometric/transforms/to_sparse_tensor.py [hunk 3]
```diff
@@ -4,39 +4,57 @@
...
+        since there exist some transforms that are only able to
         operate on :obj:`data.edge_index` for now.
 
     Args:
         attr (str, optional): The name of the attribute to add as a value to
-            the :class:`~torch_sparse.SparseTensor` object (if present).
+            the :class:`~torch_sparse.SparseTensor` or
+            :class:`torch.spase.Tensor` object (if present).
```
Comment (wsad1) [inline torch_geometric/transforms/to_sparse_tensor.py:29]:
```suggestion
            :class:`torch.sparse.Tensor` object (if present).
```

File: torch_geometric/transforms/to_sparse_tensor.py [hunk 4]
```diff
@@ -4,39 +4,57 @@
...
         fill_cache (bool, optional): If set to :obj:`False`, will not fill the
             underlying :class:`~torch_sparse.SparseTensor` cache.
+            This does not work for :class:`torch.spase.Tensor` input.
             (default: :obj:`True`)
+        backend (str, optional): It specifies the backend library to be used
+            for converting the :obj:`edge_index` to a sparse tensor. If set to
+            :obj:`'torch_sparse'`, the :obj:`edge_index` will be converted to
+            a :class:`~torch_sparse.SparseTensor` object. If set to
+            :obj:`'torch'`, the :obj:`edge_index`_index will be converted to a
+            :class:`torch.spase.Tensor` object.
```
Comment (wsad1) [inline torch_geometric/transforms/to_sparse_tensor.py:43]:
```suggestion
            :class:`torch.sparse.Tensor` object.
```

File: torch_geometric/transforms/to_sparse_tensor.py [hunk 5]
```diff
@@ -4,39 +4,57 @@
...
             (default: :obj:`True`)
         fill_cache (bool, optional): If set to :obj:`False`, will not fill the
             underlying :class:`~torch_sparse.SparseTensor` cache.
+            This does not work for :class:`torch.spase.Tensor` input.
             (default: :obj:`True`)
+        backend (str, optional): It specifies the backend library to be used
+            for converting the :obj:`edge_index` to a sparse tensor. If set to
+            :obj:`'torch_sparse'`, the :obj:`edge_index` will be converted to
+            a :class:`~torch_sparse.SparseTensor` object. If set to
+            :obj:`'torch'`, the :obj:`edge_index`_index will be converted to a
```
Comment (wsad1) [inline torch_geometric/transforms/to_sparse_tensor.py:42]:
```suggestion
            :obj:`'torch'`, the :obj:`edge_index` will be converted to a
```

**Review by wsad1 — APPROVED** [review]
Thanks!

File: torch_geometric/transforms/to_sparse_tensor.py [hunk 6]
```diff
@@ -65,19 +83,30 @@ def __call__(
...
-            if self.fill_cache:  # Pre-process some important attributes.
+            if self.fill_cache and self.backend == 'torch_sparse':
+                # Pre-process some important attributes.
                 store.adj_t.storage.rowptr()
                 store.adj_t.storage.csr2csc()
 
         return data
+
+    def __repr__(self):
+        return f'{self.__class__.__name__}(backend={self.backend})'
```
Comment (wsad1) [inline torch_geometric/transforms/to_sparse_tensor.py:112]:
```suggestion
        return (f'{self.__class__.__name__}(attr={self.attr},'
            f' remove_edge_index={self.remove_edge_index},'
            f' fill_cache={self.fill_cache}, backend={self.backend})')
```

File: torch_geometric/transforms/to_sparse_tensor.py [hunk 7]
```diff
@@ -65,19 +83,30 @@ def __call__(
...
-                store[self.attr], sparse_sizes=store.size()[::-1],
-                is_sorted=True, trust_data=True)
+            if self.backend == 'torch_sparse':
+                store.adj_t = SparseTensor(
+                    row=store.edge_index[1], col=store.edge_index[0],
+                    value=None if self.attr is None or self.attr not in store
+                    else store[self.attr], sparse_sizes=store.size()[::-1],
+                    is_sorted=True, trust_data=True)
+            else:
+                store.adj_t = to_torch_coo_tensor(
```
Comment (EdisonLeeeee) [inline torch_geometric/transforms/to_sparse_tensor.py:93]: Should we use `to_torch_csr_tensor` instead?

Comment (wsad1) [reply torch_geometric/transforms/to_sparse_tensor.py:93]: Do we want to let the user choose the layout. The default could be `csr`.

**Review by rusty1s — APPROVED** [review]
Made some small change by introducing `layout` rather than `backend`. Hope this works for you.

**Comment by EdisonLeeeee** [comment]
Super! Thank you!

**Pull request merged**